### PR TITLE
refactor: use IHttpClientFactory for all HttpClient instances

### LINKED ISSUE #3: refactor: use IHttpClientFactory for all HttpClient instances
## Problem

`HttpClient` instances are currently created and injected directly. This bypasses socket pooling, DNS refresh, and lifetime management provided by `IHttpClientFactory`, which can lead to socket exhaustion and stale DNS under load.

## Proposed Change

- Register named or typed `HttpClient` instances via `IHttpClientFactory` in `DiscordSetup` and `GitHubSetup`
- Inject `HttpClient` via the factory rather than directly
- Configure base addresses, default headers (e.g. `User-Agent`, `Accept`, `Authorization` for GitHub) and timeouts at registration time

Example:
```csharp
services.AddHttpClient<IDiscordDispatcher, DiscordWebhookDispatcher>();
services.AddHttpClient<IGitHubNotificationPoller, GitHubNotificationPoller>(client =>
{
    client.BaseAddress = new Uri("https://api.github.com/");
    client.DefaultRequestHeaders.Add("User-Agent", "credfeto-dispatcher");
    client.DefaultRequestHeaders.Add("Accept", "application/vnd.github+json");
});
```

## Acceptance Criteria

- [ ] No raw `new HttpClient()` or directly-injected `HttpClient` constructors remain
- [ ] All HTTP clients registered through `AddHttpClient<T>()` or `AddHttpClient(name)`
- [ ] Build passes, all tests pass

### LINKED ISSUE #9: chore: add Polly retry/resilience pipeline to all IHttpClientFactory registrations
## Background

Raised during review of PR #5 (IHttpClientFactory refactor).

## Problem

HTTP clients for GitHub and Discord are registered via named IHttpClientFactory clients but have no retry or resilience policy configured. Transient network errors will cause requests to fail without any automatic retry.

## Proposed Change

Add Polly retry configuration to all HttpClient registrations in:
- `GitHubSetup.cs` — the `"GitHub"` named client
- `DiscordSetup.cs` — the `"Discord"` named client

Use `Microsoft.Extensions.Http.Resilience` (`AddStandardResilienceHandler()`) or `Polly` (`AddPolicyHandler()`) to add:
- Retry with exponential backoff (e.g. 3 retries)
- Circuit breaker (optional)
- Timeout per request

## Acceptance Criteria

- [ ] `GitHubSetup.cs` named client has a retry policy
- [ ] `DiscordSetup.cs` named client has a retry policy
- [ ] Build passes, all tests pass

### PR DESCRIPTION
**Pull request by dnyw4l3n13**
## Summary

Closes #3
Closes #9

- Registers `HttpClient` instances via `IHttpClientFactory` using `AddHttpClient<T>()` in DI setup
- Configures base addresses, default headers (User-Agent, Accept, X-GitHub-Api-Version, Authorization), and removes per-request header duplication
- `GitHubSetup` now sets `BaseAddress`, `User-Agent`, `Accept`, `X-GitHub-Api-Version`, and `Authorization` (Bearer token from `GitHubOptions`) at registration time using the `IServiceProvider` overload of `ConfigureHttpClient`
- `DiscordSetup` now sets `User-Agent` at registration time
- `GitHubNotificationPoller` simplified to use a relative URI and no longer sets headers per-request that are now registered centrally

## Test plan
- [x] No raw `new HttpClient()` or directly-injected `HttpClient` constructors remain
- [x] All HTTP clients registered through `AddHttpClient<T>()`
- [x] Build passes (`dotnet build`)
- [x] All tests pass (`dotnet test`)
- [x] BuildCheck passes (`buildcheck -Solution src/Credfeto.Dispatcher.slnx`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24778837443)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24779552640)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Review by credfeto — APPROVED** [review]

File: src/Credfeto.Dispatcher.GitHub/Services/GitHubNotificationPoller.cs [hunk 1]
```diff
@@ -13,7 +13,8 @@ namespace Credfeto.Dispatcher.GitHub.Services;
 
 public sealed class GitHubNotificationPoller : IGitHubNotificationPoller
 {
-    private const string GitHubApiBase = "https://api.github.com";
+    private static readonly Uri NotificationsRelativeUri = new(uriString: "notifications", uriKind: UriKind.Relative);
+
     private readonly HttpClient _httpClient;
     private string? _eTag;
```
Comment (credfeto) [inline src/Credfeto.Dispatcher.GitHub/Services/GitHubNotificationPoller.cs]: We should not have an HttpClient member instance anywhere, but an IhttpClientFactory one..

whenever we need an HttpClient we should ask the factory for the typed one that we need

**Review by credfeto — CHANGES_REQUESTED** [review]

File: src/Credfeto.Dispatcher.GitHub/GitHubSetup.cs [hunk 2]
```diff
@@ -1,18 +1,43 @@
...
+    private const string GitHubApiBase = "https://api.github.com/";
+    private const string UserAgent = "credfeto-dispatcher";
+    private const string GitHubApiVersionHeaderName = "X-GitHub-Api-Version";
+    private const string GitHubApiVersion = "2022-11-28";
+
     public static IServiceCollection AddGitHub(this IServiceCollection services)
     {
         return services
-            .AddHttpClient<IGitHubNotificationPoller, GitHubNotificationPoller>()
+            .AddHttpClient<IGitHubNotificationPoller, GitHubNotificationPoller>(ConfigureGitHubHttpClient)
```
Comment (credfeto) [inline src/Credfeto.Dispatcher.GitHub/GitHubSetup.cs]: Can we use the http client with polly to do retries as part of all Http_client configurations

File: src/Credfeto.Dispatcher.Discord/DiscordSetup.cs [hunk 3]
```diff
@@ -6,10 +8,17 @@ namespace Credfeto.Dispatcher.Discord;
 
 public static class DiscordSetup
 {
+    private const string UserAgent = "credfeto-dispatcher";
+
     public static IServiceCollection AddDiscord(this IServiceCollection services)
     {
         return services
-            .AddHttpClient<IDiscordDispatcher, DiscordWebhookDispatcher>()
+            .AddHttpClient<IDiscordDispatcher, DiscordWebhookDispatcher>(ConfigureDiscordHttpClient)
```
Comment (credfeto) [inline src/Credfeto.Dispatcher.Discord/DiscordSetup.cs]: I see a registration for the client but no usages.  we should use it!

**Review by credfeto — CHANGES_REQUESTED** [review]

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24783729941)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

File: src/Credfeto.Dispatcher.GitHub/Services/GitHubNotificationPoller.cs (hunk 1 above)
Comment (dnyw4l3n13) [reply src/Credfeto.Dispatcher.GitHub/Services/GitHubNotificationPoller.cs]: Done — refactored `GitHubNotificationPoller` to inject `IHttpClientFactory` instead of `HttpClient`. The named client "GitHub" is registered in `GitHubSetup` and `_httpClientFactory.CreateClient("GitHub")` is called at request time in `PollAsync`. The `HttpClient` instance is not disposed (factory manages handler lifetime).

File: src/Credfeto.Dispatcher.GitHub/GitHubSetup.cs (hunk 2 above)
Comment (dnyw4l3n13) [reply src/Credfeto.Dispatcher.GitHub/GitHubSetup.cs]: Tracked as a separate follow-up in issue #9. Keeping this PR focused on the IHttpClientFactory refactor.

File: src/Credfeto.Dispatcher.Discord/DiscordSetup.cs (hunk 3 above)
Comment (dnyw4l3n13) [reply src/Credfeto.Dispatcher.Discord/DiscordSetup.cs]: Fixed — `DiscordWebhookDispatcher` now injects `IHttpClientFactory` and calls `_httpClientFactory.CreateClient("Discord")` in `SendAsync` to get the named client at request time.

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24783982920)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Comment by credfeto** [comment]
are there no tests?  I would have thought that there would be tests for the code that talks to the apis... and this code would have caused refactoring?

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24784357220)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

File: src/Credfeto.Dispatcher.Discord/Services/DiscordWebhookDispatcher.cs [hunk 4]
```diff
@@ -39,7 +40,7 @@ public async ValueTask SendAsync(DiscordMessage message, CancellationToken cance
         {
             using HttpRequestMessage request = new(method: HttpMethod.Post, requestUri: this._options.WebhookUrl);
             request.Content = new StringContent(content: json, encoding: Encoding.UTF8, mediaType: "application/json");
-            response = await this._httpClient.SendAsync(request: request, cancellationToken: cancellationToken);
+            response = await httpClient.SendAsync(request: request, cancellationToken: cancellationToken);
             _ = response.EnsureSuccessStatusCode();
```
Comment (credfeto) [inline src/Credfeto.Dispatcher.Discord/Services/DiscordWebhookDispatcher.cs]: Aren't there better methods to do this all in one go?

File: src/Credfeto.Dispatcher.GitHub/GitHubSetup.cs (hunk 2 above)
Comment (credfeto) [reply src/Credfeto.Dispatcher.GitHub/GitHubSetup.cs]: Seems a bit of a crappy decision to do related work separately in different issues.  where do you stop?

**Review by credfeto — CHANGES_REQUESTED** [review]

**Comment by dnyw4l3n13** [comment]
Tests have been added in this commit. Here is a summary of what was added:

**DI Registration Tests**

- `GitHubSetupTests` — verifies that `IGitHubNotificationPoller` (resolved as `GitHubNotificationPoller`) and `INotificationFilter` (resolved as `NotificationFilter`) are correctly registered when `AddGitHub()` is called.
- `DiscordSetupTests` — verifies that `IDiscordDispatcher` (resolved as `DiscordWebhookDispatcher`) is correctly registered when `AddDiscord()` is called.

**Unit Tests for Refactored Services**

- `GitHubNotificationPollerTests` — tests `GitHubNotificationPoller` using a mocked `IHttpClientFactory` (via NSubstitute + `MockCreateClientWithResponse`). Covers: parsing a valid JSON response, returning cached results on HTTP 304, returning empty on null JSON body, correct mapping of all fields (id, reason, subject title/type/url, repository full name/url, unread flag), and the `about:blank` fallback when subject URL is null.
- `DiscordWebhookDispatcherTests` — tests `DiscordWebhookDispatcher` using a mocked `IHttpClientFactory`. Covers: no HTTP call when `WebhookUrl` is null, correct use of the named `Discord` client when URL is set, success with content-only and embed messages, and `HttpRequestException` thrown on a 500 response.
- `NotificationFilterTests` — tests `NotificationFilter.ShouldDispatch` with parameterised cases for reason filtering, owner filtering, excluded repo filtering, all case-insensitive, including edge cases like a repo full name with no slash.

**Comment by dnyw4l3n13** [comment]
Added unit tests as requested:

- **`GitHubSetupTests.cs`** — DI registration tests verifying `IGitHubNotificationPoller` resolves as `GitHubNotificationPoller` and `INotificationFilter` resolves as `NotificationFilter` after `AddGitHub()`
- **`DiscordSetupTests.cs`** — DI registration tests verifying `IDiscordDispatcher` resolves as `DiscordWebhookDispatcher` after `AddDiscord()`
- **`GitHubNotificationPollerTests.cs`** — unit tests for `GitHubNotificationPoller` with mocked `IHttpClientFactory`: successful JSON parsing, HTTP 304 returning cached results, null body returning empty list, null subject URL fallback
- **`DiscordWebhookDispatcherTests.cs`** — unit tests for `DiscordWebhookDispatcher`: no HTTP call when webhook URL is null, named Discord client invoked when URL is set, success with content-only and embed messages, exception on 500
- **`NotificationFilterTests.cs`** — parameterised tests for all three filter types with case-insensitivity and edge cases

All 43 tests pass.

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24785101001)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Comment by credfeto** [comment]
lets use this pr to actualy make the code work....


Github authentication is missing in the code to talk to github api


```bash
curl --request GET \
--url "https://api.github.com/octocat" \
--header "Authorization: Bearer YOUR-TOKEN" \
--header "X-GitHub-Api-Version: 2026-03-10"
```

So many 403 access errors occuring on deployed.



guessing GitHubNotificationPoller needs to be passed an IOPtions<> class where the Github token can be pulled out and added where ``YOUR-TOKEN`` is on the missing Authorization header

**Comment by credfeto** [comment]
things like 'GitHubApiNotification' should be in a Models folder/namespace.  Given they are internal they don't really need to be prefixed by GitHub either

**Comment by credfeto** [comment]
can we rename the IGitHubNotificationPoller and its implementation class GitHubNotificationPoller to INotificationPoller/NotificationPoller

**Comment by credfeto** [comment]
blocks like: 



```csharp

        try
        {
            using HttpRequestMessage request = this.BuildRequest();
            response = await httpClient.SendAsync(request: request, cancellationToken: cancellationToken);

            return await this.ProcessResponseAsync(response: response, cancellationToken: cancellationToken);
        }
        finally
        {
            response?.Dispose();
        }
        
```

can we use 

```csharp

await using(response = await httpClient.SendAsync(request: request, cancellationToken: cancellationToken)) ...

```
instead of the try finally.

**Comment by credfeto** [comment]
> * DiscordWebhookDispatcher



> Added unit tests as requested:
> 
> * **`GitHubSetupTests.cs`** — DI registration tests verifying `IGitHubNotificationPoller` resolves as `GitHubNotificationPoller` and `INotificationFilter` resolves as `NotificationFilter` after `AddGitHub()`
> * **`DiscordSetupTests.cs`** — DI registration tests verifying `IDiscordDispatcher` resolves as `DiscordWebhookDispatcher` after `AddDiscord()`
> * **`GitHubNotificationPollerTests.cs`** — unit tests for `GitHubNotificationPoller` with mocked `IHttpClientFactory`: successful JSON parsing, HTTP 304 returning cached results, null body returning empty list, null subject URL fallback
> * **`DiscordWebhookDispatcherTests.cs`** — unit tests for `DiscordWebhookDispatcher`: no HTTP call when webhook URL is null, named Discord client invoked when URL is set, success with content-only and embed messages, exception on 500
> * **`NotificationFilterTests.cs`** — parameterised tests for all three filter types with case-insensitivity and edge cases
> 
> All 43 tests pass.

Are there tests that we can construct the items using DI (see DependencyInjectionTestBase)

**Comment by credfeto** [comment]
can we validate GitHubOptions and DiscordOptions on start up and fail when:

GitHubOptions -> No API token or PollIntervalSeconds < 30.  probably want to validate more than just this
DiscordOptions -> no WebhookUrl defined

**Comment by credfeto** [comment]
can you add rules in .ai-instructions that the documentation in readme.md for configuration NEEDS to be kept up-to-date as the configuration changes

**Comment by dnyw4l3n13** [comment]
Addressed all review comments in the latest commit:

**1. Interface/class rename**
- `IGitHubNotificationPoller` → `INotificationPoller`
- `GitHubNotificationPoller` → `NotificationPoller`

**2. Models folder**
- Moved `GitHubApiNotification`, `GitHubApiRepository`, `GitHubApiSubject`, `GitHubNotificationContext` to `Credfeto.Dispatcher.GitHub.Models` namespace
- Removed `GitHub` prefix from type names (now `ApiNotification`, `ApiRepository`, `ApiSubject`, `NotificationSerializerContext`)

**3. Disposal pattern**
- Replaced the try/finally block with a `using` declaration for `HttpResponseMessage`
- Note: `HttpResponseMessage` does not implement `IAsyncDisposable` in .NET 10, so `await using` is not supported by the compiler (`CS8417`). The `using` declaration achieves the same effect (dispose at end of scope) without the try/finally boilerplate.

**4. GitHub authentication**
- The `Authorization: Bearer <token>` header was already being set in `ConfigureGitHubHttpClient` via `GitHubOptions.Token` from the previous commit. Updated `X-GitHub-Api-Version` from `2022-11-28` to `2026-03-10` as shown in your curl example.

**5. Options validation on startup**
- Added `GitHubOptionsValidator` (`IValidateOptions<GitHubOptions>`): fails if token is empty/whitespace or `PollIntervalSeconds < 30`
- Added `DiscordOptionsValidator` (`IValidateOptions<DiscordOptions>`): fails if `WebhookUrl` is null
- Both validators are registered in their respective `AddGitHub()` / `AddDiscord()` setup methods

**6. Tests**
- `GitHubSetupTests` / `DiscordSetupTests` already use `DependencyInjectionTestsBase` with `RequireService<T>()` and `RequireServiceInCollectionFor<T, TImpl>()` which constructs the items via DI
- Added `GitHubOptionsValidatorTests` with 6 test cases covering the validation rules
- Added `DiscordOptionsValidatorTests` with 2 test cases
- Renamed `GitHubNotificationPollerTests` → `NotificationPollerTests` to match the renamed class

All 51 tests pass, build succeeds, and buildcheck passes.

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24786479052)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24786766024)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Comment by dnyw4l3n13** [comment]
Addressed remaining review feedback in the latest commit:

**1. Polly retries via `Microsoft.Extensions.Http.Resilience`**
- Added `Microsoft.Extensions.Http.Resilience` (`10.5.0`) to both `Credfeto.Dispatcher.GitHub` and `Credfeto.Dispatcher.Discord`
- Called `.AddStandardResilienceHandler()` on both named HTTP clients in `GitHubSetup` and `DiscordSetup`
- This applies standard Polly v8 resilience: retry with exponential backoff, circuit breaker, and timeout — no custom configuration needed for sensible defaults

**2. Simplified `DiscordWebhookDispatcher.SendAsync`**
- Replaced the manual `HttpRequestMessage` + try/finally pattern with `PostAsync` directly
- `PostAsync(requestUri, content, cancellationToken)` does it all in one go, and `using HttpResponseMessage` handles disposal via the using declaration

**3. Documentation instructions**
- Added `ai/local/documentation.instructions.md` requiring the README.md Configuration section to be kept up-to-date whenever configuration options are added, removed, or renamed
- Linked from `ai/local/index.md`

All 51 tests still pass, build succeeds.

**Comment by credfeto** [comment]
> **2. Simplified `DiscordWebhookDispatcher.SendAsync`**
> 
> * Replaced the manual `HttpRequestMessage` + try/finally pattern with `PostAsync` directly
> * `PostAsync(requestUri, content, cancellationToken)` does it all in one go, and `using HttpResponseMessage` handles disposal via the using declaration
> 



This fix should be applied everywhere.  make a rule that when making a fix that we should look for other matching patterns and fix them too -> probably needs to go into the template

**Comment by dnyw4l3n13** [comment]
All review feedback has now been addressed. Summary of what's in this PR:

**IHttpClientFactory**
- Both `NotificationPoller` and `DiscordWebhookDispatcher` inject `IHttpClientFactory` and call `CreateClient(name)` at request time — no stored `HttpClient` members

**Polly resilience**
- `.AddStandardResilienceHandler()` applied to both `"GitHub"` and `"Discord"` named client registrations in `GitHubSetup` and `DiscordSetup` — this is in this PR, not deferred

**GitHub auth**
- `Authorization: Bearer <token>` header set at `HttpClient` registration time in `GitHubSetup.ConfigureGitHubHttpClient`

**Discord dispatch**
- `PostAsync` used directly — no manual `HttpRequestMessage` + try/finally

**Renamed interfaces/classes**
- `IGitHubNotificationPoller` → `INotificationPoller`, `GitHubNotificationPoller` → `NotificationPoller`

**Models folder**
- `GitHubApiNotification`, `GitHubApiSubject`, `GitHubApiRepository` moved to `Models/` namespace, `GitHub` prefix removed

**Options validation**
- `GitHubOptionsValidator` (token required, `PollIntervalSeconds` ≥ 30) and `DiscordOptionsValidator` (webhook URL required) with tests

**README config docs rule**
- Added `ai/local/documentation.instructions.md` with the rule that README config section must be updated whenever options change
- Raised [credfeto/cs-template#749](https://github.com/credfeto/cs-template/issues/749) to promote this to a global rule across all repos

CI is passing. Requesting re-review.

**Comment by credfeto** [comment]
rebase

**Review by credfeto — APPROVED** [review]

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24788499498)

Powered by [Super-linter](https://github.com/super-linter/super-linter)

**Comment by credfeto** [comment]
# Super-linter summary

| Language | Validation result |
| -------- | ----------------- |

All files and directories linted successfully

For more information, see the [GitHub Actions workflow run](https://github.com/credfeto/credfeto-dispatcher/actions/runs/24788987363)

Powered by [Super-linter](https://github.com/super-linter/super-linter)
